### ISSUE TITLE
Script doesn't work around flickr's faulty paging 

### ISSUE THREAD

**Issue by alexhunsley** [markdoliner/flickrmirrorer#21]
Yeah, I know this is flickr API's problem, not flickrmirrorer's, but flickr apparently say "you have to work around it" :(

More info:

https://www.flickr.com/groups/51035612836@N01/discuss/72157685797001754/

I've discovered when using flickrmirrorer that I get incomplete photostream downloaded, and this seems to be the cause.

Illustration of problem: when requesting JSON pages of results, 500 results per page, you can get back <500 results in a JSON page. flickrmirrorer doesn't detect this issue and just assumes all is fine and iterates through however many results it gets back in a page. The result is that you end up with album and collection symlinks to non-existent files.

Looking at making a fix.

**Comment by markdoliner** [comment]
That's interesting, I haven't hard of this problem before. What do you think the right fix is? Maybe: `if num photos in result < NUM_PHOTOS_PER_BATCH and current_page < num_pages then retry the API request and hope it includes all the photos the next time around`? And it seems like there might be no way to detect this problem for the last page. Unless maybe the API tells us how many photos the user has total? Not sure.

I'll add logging to my local script and see if this situation happens for me.

**Comment by alexhunsley** [comment]
I was going with the retry approach, pretty much as you outlined, but on the day I was running the script it repeatedly give me back too few results on exactly the same page each time! e.g. I was getting only 497 results on page 7, every time.

In the end I went with an approach of making the script bomb out and report a problem to the user -- at least then they know there was a problem, and to try running the script again the next day.

However, then I found the other issue: there can be photos referenced from albums that don't appear in your photostream! So I also made a change to make the 'download albums' and 'download photos not in an album' part actually download the photos it found, rather than just creating symlinks to them (which assumes the photostream downloads everything).

I'll review and tidy up where my fork of your script is and get back to you soon.

In summary, I think I know why so many flickr programs (Bulkr etc) seemed to be flaky - the underlying flickr API can be quite flaky!

**Comment by alexhunsley** [comment]
Update: it's a while since I've done anything on this, but I just pushed all changes to my fork (https://github.com/alexhunsley/flickrmirrorer.git). It's certainly not in any usefully merge-able state, but you might be interested in the API rate limiting I added along the way (e.g. first commit for that is a57f498)
